collectors/google_pubsub/pubsub_collector.go

```diff
@@ -36,19 +60,143 @@ func (collector *PubSubCollector) HandleMessage(message *pubsub.Message) error {
 	return result.Err
 }
 
+func (collector *PubSubCollector) PollForMessages() {
+	go collector.timeout()
+	for {
+		select {
+		case <-collector.ChannelToInitiatePollRequest:
+			if !collector.BaseCollector.Sleeping {
+				messages, err := collector.GetMessages()
+				if err != nil {
+					log.Println("Error receiving messages - retrying")
+					collector.ProcessExponentialBackoff(err)
+				}
+				for _, msg := range messages {
+					result := collector.BusinessProcessor(msg.MessageBody)
+					if result.Err != nil {
+						log.Printf("Buisness Logic error - %s", result.Err)
+						msg.Retries++
+
+						//TODO - you are here, need to figure out how to make this stop posting to original topic
+						if msg.Retries > collector.MaxRetries {
+							collector.PublishMessage(&msg, 0, true)
+						}
+						collector.PublishMessage(&msg, 0, false)
+					}
+				}
+			}
+			go collector.timeout()
+		case <-collector.ChannelDone:
+			return
+		}
+	}
+}
+
+func (collector *PubSubCollector) timeout() {
+	time.Sleep(time.Duration(collector.PollingPeriod) * time.Millisecond)
+	collector.ChannelToInitiatePollRequest <- true
+}
+
 func (collector *PubSubCollector) GetMessages() ([]collectors.MessageWrapper, error) {
-	var err error
+	log.Printf("Get messages start")
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// There's a built in deadline on the subscription that will orphan the message in the subscription.
+	subscription := collector.PubSubClient.Subscription(collector.SourceSubscription)
 	messages := make([]collectors.MessageWrapper, 0)
 
+	// Track number of messages received and timer to cancel this function when needed
+	found := make(chan int, 1)
+	received := 0
+	timer := time.NewTimer(5 * time.Second)
+	go func() {
+		for {
+			select {
+			case <- found:
+				received++
+				if received >= collector.PollingLimit {
+					log.Printf("Recieved %s messages\n", collector.PollingLimit)
+					cancel()
+				}
+			case <- timer.C:
+				log.Println("Timer fired")
+				cancel()
+			}
+		}
+	}()
+
+	// TODO - see if I need this
+	var mu sync.Mutex
+	err := subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		log.Printf("Got message: %q\n", string(msg.Data))
+		// If value is not present in map we send an empty string in here, which will result in a 0 int.
+		retries, _ := strconv.Atoi(msg.Attributes["retries"])
+
+		messages = append(messages, collectors.MessageWrapper{
+			MessageBody: msg.Data,
+			Retries:     retries,
+			Retry:       false,
+			Fatal:       false,
+			Message:     msg,
+		})
+		msg.Ack()
```

Comment:
Remember we have a method to call to Ack() messages..  If you ack here the BaseCollector has no idea and cannot retry or anything with this message

Reply:
Ah, so I see the BaseCollector logic now.  I wired that in, but now I'm getting a weird "bad access: nil dereference" error.   When it tries to re-publish.  The only thing I can think of is that it can't find my PublishMessage method.  Though I have no idea why that would be - might need eyes from the goru himself with a little paired go-age? 😄

Reply:
lol